File: UI/markup/js/hardware.js
// Copyright (c) 2015 David Wilson
// Icarus is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Icarus is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Icarus.  If not, see <http://www.gnu.org/licenses/>.

var Hardware = function(ajax, urls) {
	 this.ajax = ajax;
	 this.urls = urls;
};

/**
 * Get the details of the item of hardware currently being viewed.
 * 
 * @return {object} The details of the current item of hardware as an object.
 */
Hardware.prototype.getHardwareNoId = function() {
	 return {
        name: $("#name").val(),
        platform: $("#platform").val(),
        numcopies: $("#numcopies").val(),
        numboxed: $("#numboxed").val(),
        notes: $("#notes").val(),
		  hardwaretype: $("#hardwaretype").val()		  
    };
};

/**
 * Delete the item of hardware being currently viewed. This is done using ajax and after the 
 * item of hardware has been saved the user will be redirected to the All Hardware page.
 */
Hardware.prototype.deleteHardware = function() {

	 var def = $.Deferred();

	 this.ajax.ajaxDelete(urls.deletehardware, this.ajax.getIdJson())
		  .done(function(r) { def.resolve(r); })
		  .fail(function(r) { def.reject(r); });

	 return def;
};

/**
 * Add a new item of hardware. This is done using ajax and after the
 * item of hardware has been saved the user will be redirected to the All Hardware page.
 */
Hardware.prototype.addHardware = function(h, a) {
	 var def = $.Deferred();

	 var j = h.getHardwareNoId();
	 
	 var validationResult = h.validateSaveHardwareJson(j);

	 if (validationResult.result === 'fail') {
		  def.reject({'fields': validationResult.fields});
		  return def;
	 }

	 a.ajaxSave(urls.savehardware, j)
		  .done(function(r) { 
				def.resolve(r); 
		  })
		  .fail(function(r) { 
				def.reject(r);
		  });		  

	 return def;
};

/**
 * Update an item of hardware. This is done using ajax and after the
 * item of hardware has been saved the user will be redirected to the All Hardware page.
 */
Hardware.prototype.updateHardware = function(h, a) {
	 var def = $.Deferred();

	 var j = h.getHardwareNoId();
	 j.id = a.getIdJson().id;

	 var validationResult = h.validateSaveHardwareJson(j);

	 if (validationResult.result === 'fail') {
		  def.reject({'fields': validationResult.fields});
		  return def;
	 }

	 a.ajaxSave(urls.updatehardware, j)
		  .done(function(r) { 
				def.resolve(r); 
		  })
		  .fail(function(r) { 
				def.reject(r); 
		  });

	 return def;
};

Hardware.prototype.validateSaveHardwareJson = function(j) {
	 var requiredFields = ['name', 'numowned', 'numboxed'];

	 var missingFields = requiredFields.filter(function(f) {
		  return j[f] === '';
	 });

	 return {
		  'result': missingFields.length === 0? 'success': 'fail',
		  'fields': missingFields
	 };
};

/**
 * Sort the list of items of hardware on screen.
 *
 * @param {field} The field to sort by
 */
Hardware.prototype.sortHardware = function(field) {
    var hdnSort = $('#hwsortfield');
    var hdnDir = $('#hwsortdir');
    var hdnRows = $('#gamerows');

    var oldSortDir = hdnDir.val();
    var newSortDir = "asc";
    var numRows = hdnRows.val() === null ? 999999 : hdnRows.val();

    if (hdnSort.val() == field) newSortDir = toggleSortDirection(oldSortDir);

    hdnSort.val(field);
    hdnDir.val(newSortDir);
	
	 var loadData = {
        field: field,
        sortdir: newSortDir,
        numrows: numRows
    };

	 this.ajax.loadAjax('#hardware', urls.sorthardware, loadData, this.initSorting);
};

Hardware.prototype.initSorting = function(hardware, status) {
	 if (status!=='1') hardware = new Hardware(new Ajax(), urls);

	 $('a.sorthardwarename').on('click', function(e) {
	 	  hardware.sortHardware('name');
	 });

	 $('a.sorthardwareplatform').on('click', function(e) {
	 	  hardware.sortHardware('platform');
	 });

	 $('a.sorthardwarenumowned').on('click', function(e) {
	 	  hardware.sortHardware('numowned');
	 });
};

$(function() {
	 var a = new Ajax();
	 var h = new Hardware(a, urls);
	 
	 h.initSorting(h, '1');

	 $('input.saveButton').on('click', function(e) {

		  var updateHardware = function() {
				if (document.location.pathname === urls.edithardware) return h.updateHardware;
				if (document.location.pathname === urls.addhardware) return h.addHardware;
		  };

		  var saveHardwareFunc = updateHardware();
		  
		  var button = $('input.saveButton');
		  var loadingGifClass = 'save-hardware-loading-gif';
		  var inputs = $('input, select, textarea');		  
		  
		  doLoading(button, loadingGifClass, inputs);

		  e.preventDefault();

		  saveHardwareFunc(h, a)
				.done(function() { 
					 document.location = urls.allhardware; 
				})
				.fail(function(r) {
					 if (r.fields) showFailure(r.fields, 'hardware');
				})
				.always(function() {
					 finishedLoading(loadingGifClass, inputs);
				});
	 });

	 $('a.yesDelete').on('click', function(e) {
		  
		  var link = $('a.yesDelete');
		  var loadingGifClass = 'delete-hardware-loading-gif';
		  var inputs = $('input, select, textarea');
		  
		  e.preventDefault();

		  doLoading(link, loadingGifClass, inputs);
		  h.deleteHardware()
		  		.done(function() { 
					 document.location = urls.allhardware; 
				})
				.always(function() {
					 finishedLoading(loadingGifClass, inputs);
				});
		  return false;
	 });
});

File: hardware.py
"""Represents an item of hardware"""
# Copyright (c) 2015, 2026 David Wilson
# Icarus is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# Icarus is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with Icarus.  If not, see <http://www.gnu.org/licenses/>.

from dataclasses import dataclass, field
import functools as ft
import json
from typing import Any

@dataclass
class Hardware():
    """Represents an item of hardware"""
    id: str = field(default="")
    hardware_type: str = field(default="")
    name: str = field(default="")
    platform: str = field(default="")
    num_owned: str = field(default="")
    num_boxed: str = field(default="")
    notes: str = field(default="")
    user_id: str = field(default="")

    @staticmethod
    def from_dict(dictionary: dict[str, Any]) -> "Hardware":
        """Initialises Hardware object from a dictionary.
        :param dictionary: A dictionary containing the following keys:
                            * name
                            * platform
                            * numcopies
                            * numboxed
                            * notes
                            * hardware_type
        :returns: A Hardware object with its properties properly initialised. Any missing keys 
                    from dictionary will cause the object to have that properly initialised as 
                    its default.
        """

        # hardware.attr, d.key
        mappings = {"id": "id",
                    "name": "name",
                    "platform": "platform",
                    "num_owned": "numcopies",
                    "num_boxed": "numboxed",
                    "notes": "notes",
                    "user_id": "userid",
                    "hardware_type": "hardwaretype"}

        return Hardware._from_dict(dictionary, mappings)

    #TODO: Can probably clean this dict stuff up between the domain objects
    @staticmethod
    def _from_dict(dictionary: dict[str, Any], mappings: dict[str, str]) -> "Hardware":
        hardware = Hardware()

        set_attr = ft.partial(setattr, hardware)
        get_attr = ft.partial(getattr, hardware)

        def dict_get(x: tuple[str, str]) -> Any:
            return dictionary.get(x[0], get_attr(x[1]))

        list(map(lambda m: set_attr(m, dict_get((mappings[m],m))), mappings))
        return hardware

    def to_json(self) -> str:
        """Convert this Hardware object to a JSON string"""

        attrs = ["hardware_type", "name", "platform", "num_owned", "num_boxed", "notes"]
        result = {}

        result["id"] = str(self.id)

        for a in attrs:
            result[a] = getattr(self, a)

        return json.dumps(result)
